### PR TITLE
[FLINK-34220][table-runtime] Introduce BufferBundle for minibatch join

### PR DESCRIPTION
**Pull request by xishuaidelin**
## What is the purpose of the change

This pull request introduces a new buffer bundle to support minibatch join. 

## Brief change log

Added a class called BufferBundle.
  -  InputSideHasNoUniqueKeyBundle for input that has no uniqueKey.
  -  InputSideHasUniqueKeyBundle for input that has uniqueKey.
  -  JoinKeyContainsUniqueKeyBundle for input whose uniqueKey is contained by joinKey.

## Verifying this change

Add BufferBundleTest  to verify the logic in BufferBundle.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

**Comment by xishuaidelin** [comment]
@flinkbot run azure

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/bundle/InputSideHasNoUniqueKeyBundle.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    private int actualSize;
+
+    public InputSideHasNoUniqueKeyBundle() {
+        this.bundle = new HashMap<>();
+        this.count = 0;
+        this.actualSize = 0;
+    }
+
+    @Override
+    public boolean isEmpty() {
```
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/bundle/InputSideHasNoUniqueKeyBundle.java]: @lsyldliu  yes，I hava extracted them. clear method includes bundle, so it is left in the subclass.

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/bundle/BufferBundle.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A {@link BufferBundle} is a bundle to buffer the input records in memory and fold data based on
+ * specified pattern to reduce state access. The bundle is used in MiniBatchStreamingJoinOperator.
+ * The structure of the bundle varies depending on the {@link JoinInputSideSpec}.
+ */
+@Internal
+public interface BufferBundle {
```
[the remark this answers is not in the document]
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/bundle/BufferBundle.java]: @lsyldliu  good idea!

**Comment by xishuaidelin** [comment]
@flinkbot run azure

**Comment by xishuaidelin** [comment]
@flinkbot run azure

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/bundle/InputSideHasNoUniqueKeyBundle.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+
+/** For the case that input has no uniqueKey. */
+public class InputSideHasNoUniqueKeyBundle extends BufferBundle<Map<Integer, List<RowData>>> {
+
+    public InputSideHasNoUniqueKeyBundle() {
```
Comment (lsyldliu) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/bundle/InputSideHasNoUniqueKeyBundle.java:36]: Why do we need to implement an empty construct method?

**Comment by xishuaidelin** [comment]
@flinkbot run azure

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/bundle/InputSideHasNoUniqueKeyBundle.java (hunk 3 above)
Comment (xishuaidelin) [reply flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/bundle/InputSideHasNoUniqueKeyBundle.java]: Yes, It is not necessary. I hava removed them.

**Comment by xishuaidelin** [comment]
@flinkbot run azure

**Review by lsyldliu — APPROVED** [review]
LGTM

**Pull request closed without merging**